dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/sms/listener/EnrollmentSMSListener.java

```diff
@@ -69,28 +70,37 @@ protected void postProcess(IncomingSms sms, SMSSubmission submission) {
             throw new SMSParserException(NO_OU_FOR_USER);
         }
 
-        TrackedEntityInstance entityInstance = new TrackedEntityInstance();
-        entityInstance.setOrganisationUnit(ou);
-        entityInstance.setTrackedEntityType(entityType);
+        TrackedEntityInstance entityInstance;
+        boolean existsOnServer = teiService.trackedEntityInstanceExists(teiUID);
 
-        Set<TrackedEntityAttributeValue> attributeValues = subm.getValues()
-                .stream()
-                .filter(this::isValidAttribute)
-                .map(v -> createTrackedEntityValue(v, entityInstance))
-                .collect(Collectors.toSet());
+        if (existsOnServer) {
+            entityInstance = teiService.getTrackedEntityInstance(teiUID);
+        } else {
+            entityInstance = new TrackedEntityInstance();
+            entityInstance.setUid(teiUID);
+            entityInstance.setOrganisationUnit(ou);
+            entityInstance.setTrackedEntityType(entityType);
+        }
+
+        Set<TrackedEntityAttributeValue> attributeValues = getAttributeValues(subm, entityInstance);
 
         if (attributeValues.isEmpty()) {
             sendFeedback(NO_TRACKED_ATTRIBUTES_FOUND, sender, WARNING);
             // TODO: Should we throw an error here or continue as is?
         }
 
         entityInstance.setTrackedEntityAttributeValues(attributeValues);
-        int teiID = teiService.addTrackedEntityInstance(entityInstance);
 
-        TrackedEntityInstance tei = teiService.getTrackedEntityInstance(teiID);
+        if (existsOnServer) {
+            teiService.updateTrackedEntityInstance(entityInstance);
+        } else {
+            teiService.addTrackedEntityInstance(entityInstance);
+        }
+
+        TrackedEntityInstance tei = teiService.getTrackedEntityInstance(teiUID);
```

Comment:
Can you quickly comment on why this is needed? Why can't we use the existing entityInstance we've created/updated? Whatever the reason, it's probably a good idea to explain it in a comment here.